app/src/main/java/com/marverenic/music/player/MusicPlayer.java

```diff
@@ -679,6 +680,7 @@ public void pause() {
         if (isPlaying()) {
             mMediaPlayer.pause();
             updateNowPlaying();
+            new SlsInteraction(mContext).sendPause(getNowPlaying());
```

Comment:
I'm a bit concerned about putting this logic here because, although simple, it adds yet another responsibility to `MusicPlayer`. The main goal of the class is to provide a high-level implementation of a music player – so things like shuffling, repeat modes, and integration with Android audio frameworks like audio focus and media session.

Scrobbling to Last.fm isn't _really_ critical to the behavior of a music player – it's at a higher level than this class, and if you're not careful about this sort of thing then all of a sudden one file ends up doing everything related to music playback in a very fragile and unmaintainable way (Take a look at the [single responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle) if you're interested in more info about this).

The way I'd go about this is to make a new interface (maybe called `MusicPlayerExtension`?) with various callbacks for player events like `onPlay`, `onPause`, etc. Then you can give `MusicPlayer` an `addExtension` and `removeExtension` method that adds/removes a callback. With this in place, all of the behavior to send Last.fm events can live in one file. (Also, if you do make a `MusicPlayerExtension` interface, it's probably better to make it an abstract class with empty method bodies so that extensions don't have to implement every single callback if they don't need them).

Reply:
> I'm a bit concerned

I was concerned about this too but there was nothing (except the `OnPlaybackChangeListener` which didn't suit for this). I agree that making such an interface is better idea.